provider/openstack/provider.go

```diff
@@ -719,8 +721,19 @@ func newCredentials(spec environs.CloudSpec) (identity.Credentials, identity.Aut
 		cred.ProjectDomain = credAttrs[CredAttrProjectDomainName]
 		cred.UserDomain = credAttrs[CredAttrUserDomainName]
 		cred.Domain = credAttrs[CredAttrDomainName]
-		authMode = identity.AuthUserPass
-		if cred.Domain != "" || cred.UserDomain != "" || cred.ProjectDomain != "" {
+		if credAttrs[CredAttrVersion] != "" {
+			version, err := strconv.Atoi(credAttrs[CredAttrVersion])
+			if err == nil {
+				if version < 3 {
+					authMode = identity.AuthUserPass
+				} else {
+					authMode = identity.AuthUserPassV3
+				}
+				cred.Version = version
+			} else {
+				logger.Debugf("cred.Version is not valid int type.")
```

Comment:
This will result in authMode not being set.
You need to add back the line: authMode = identity.AuthUserPass
There needs to be several unit tests for this code to pick up errors like this